src/hardware/imf.cpp

```diff
@@ -0,0 +1,13422 @@
+
+	std::string toString() const
+	{
+		return m_noteToString[value % 12] + "-" + std::to_string(value / 12);
+	}
+};
+#pragma pack(pop)
+static_assert(sizeof(Note) == 1, "Note needs to be 1 in size!");
+
+constexpr bool operator==(const Note& a, const Note& b)
+{
+	return a.value == b.value;
+}
+
+#pragma pack(push, 1)
+struct Fraction {
+	uint8_t value = 0;
+
+	constexpr Fraction() = default;
+	constexpr Fraction(const uint8_t v) : value(v) {}
+};
+#pragma pack(pop)
+static_assert(sizeof(Fraction) == 1, "Fraction needs to be 1 in size!");
+
+constexpr bool operator==(const Fraction& a, const Fraction& b)
+{
+	return a.value == b.value;
+}
+static Fraction ZERO_FRACTION(0);
+
+#pragma pack(push, 1)
+struct FractionalNote {
+	Fraction fraction = {};
+	Note note         = {};
+
+	constexpr FractionalNote() = default;
+	constexpr FractionalNote(const Note& nn, const Fraction& nf)
+	        : fraction(nf),
+	          note(nn)
+	{}
+	constexpr uint16_t getuint16_t() const noexcept
+	{
+		return note.value << 8 | fraction.value;
+	}
+};
+#pragma pack(pop)
+static_assert(sizeof(FractionalNote) == 2, "FractionalNote needs to be  in size!");
+
+constexpr std::pair<uint8_t, uint8_t> split_uint16_t(const uint16_t value) noexcept
+{
+	const auto upper = static_cast<uint8_t>((value >> 8) & 0xFF);
+	const auto lower = static_cast<uint8_t>(value & 0xFF);
+	return {upper, lower};
+}
+
+static constexpr FractionalNote to_fractional_note(const uint16_t value) noexcept
+{
+	const auto [note, fraction] = split_uint16_t(value);
+	return {Note(note), Fraction(fraction)};
+}
+
+
+
+constexpr bool operator==(const FractionalNote& a, const FractionalNote& b) noexcept
+{
+	return a.note == b.note && a.fraction == b.fraction;
+}
+constexpr bool operator!=(const FractionalNote& a, const FractionalNote& b) noexcept
+{
+	return !(a == b);
+}
+constexpr FractionalNote operator-(const FractionalNote& a,
+                                   const FractionalNote& b) noexcept
+{
+	const auto val = check_cast<uint16_t>(a.getuint16_t() - b.getuint16_t());
+	return to_fractional_note(val);
+}
+
+constexpr FractionalNote operator+(const FractionalNote& a,
+                                   const FractionalNote& b) noexcept
+{
+	// 16-bit wrap-around is expected and normal
+	const auto val = static_cast<uint16_t>(a.getuint16_t() + b.getuint16_t());
+	return to_fractional_note(val);
+}
+
+#pragma pack(push, 1)
+struct KeyVelocity {
+	uint8_t value = 0;
+
+	KeyVelocity() : value(0) {}
+	explicit KeyVelocity(uint8_t v) : value(v) {}
+};
+#pragma pack(pop)
+static_assert(sizeof(KeyVelocity) == 1, "KeyVelocity needs to be 1 in size!");
+
+#pragma pack(push, 1)
+struct Duration {
+	uint16_t value = 0;
+
+	Duration() : value(0) {}
+	explicit Duration(uint16_t v) : value(v) {}
+	inline bool isEmpty() const
+	{
+		return (value & 0x8000) != 0;
+	}
+	inline bool isNotEmpty()
+	{
+		return !isEmpty();
+	}
+	inline void setEmpty()
+	{
+		value |= 0x8000;
```

Comment:
Extract duplicate 0x8000 constant and give it some name?

Reply:
(not resolving because they  map to assembly literals)